torch/csrc/autograd/engine.cpp

```diff
@@ -858,33 +927,45 @@ bool Engine::is_checkpoint_valid() {
   return checkpoint_valid;
 }
 
-size_t Engine::ready_queue_size(at::Device device) {
+size_t Engine::ready_queue_size(const std::shared_ptr<GraphTask>& graph_task, at::Device device) {
   if (ready_queues_.empty()) {
     // The vector ready_queues_ is initialized in start_threads, but this method
     // can be called before start_threads. Adding this check to avoid index
     // out of bound error.
     return 0;
   }
-  return ready_queue(device).size();
+  return ready_queue(graph_task, device).size();
 }
 
-auto Engine::ready_queue(at::Device device) -> ReadyQueue& {
-  // See Note [Allocating GPUs to autograd threads]
+auto Engine::ready_queue(const std::shared_ptr<GraphTask>& graph_task, at::Device device) -> ReadyQueue& {
   if (device.type() == at::kCPU) {
-    return *ready_queues_.at(0);
+    // return the cpu ready queue memorized in GraphTask
+    TORCH_INTERNAL_ASSERT(graph_task->cpu_ready_queue_ != nullptr);
+    return *graph_task->cpu_ready_queue_;
   } else {
-    return *ready_queues_.at(device.index() + 1);
+    // See Note [Allocating GPUs to autograd threads]
+    return *ready_queues_.at(device.index());
   }
 }
 
 // See Note [Allocating GPUs to autograd threads]
 // NB: This would become obsolete if we truly allocated a CPU thread
 // per device, rather than colocate.
-auto Engine::ready_queue_by_index(int device_index) -> ReadyQueue& {
-  return *ready_queues_.at(device_index + 1);
+auto Engine::ready_queue_by_index(const std::shared_ptr<GraphTask>& graph_task, int device_index) -> ReadyQueue& {
+  if (device_index == CPU_DEVICE) {
+    if (graph_task == nullptr) {
```

Comment:
Can `graph_task` ever actually be null? If not, I'd prefer if we just internal assert this case away